### PR TITLE
[precompile] Tests for torch.compiler.precompile (eager backend)

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #190226
* #189622
* #189165
* #189621
* __->__ #189629

> **Split for PR size / reviewability.** These tests were split out from the
> implementation (**#189620**, directly below in the stack) so each PR stays under
> PyTorch's ~2000-line PR sanity-check (the CI lint that flags oversized diffs).
> This PR contains **only** `test/test_precompile.py`; it depends on and is meant
> to land together with #189620.

The test suite for torch.compiler.precompile. Covers the eager-backend calling
convention and every enforced invariant of Note [precompile programming model]:

- capture: constant-tensor / global / unregistered-attr rejection, control-flow
  and effectful-op rejection, decompositions table threading
- calling convention: param/buffer lifting and interning (tied weights, multiple
  modules), pytree in/out spec handling and unserializable-spec degradation
- runtime validation: structural mismatch, shape / dtype / device mismatch,
  code_hash pairing, corrupt / missing cache envelope fallback
- training: backward harvest matches eager (accumulation, frozen params,
  pre-existing grads), buffer-requiring-grad rejection
- public API surface (torch.compiler.precompile)

Numeric-correctness tests are device-generic via instantiate_device_type_tests.

Test Plan:

```
python test/test_precompile.py
```

Authored with an AI assistant.

### CODE REVIEW

File: test/test_precompile.py [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+        # load() surfaces a clear PrecompileError (not a raw SyntaxError) when
+        # python_code is not valid Python.
+        buf = io.BytesIO()
+        torch.save({"artifact": None}, buf)
+        with self.assertRaisesRegex(PrecompileError, "not valid Python"):
+            torch.compiler.precompile.load("def (:::", buf.getvalue())
+
+    def test_same_count_different_structure_rejected_eager(self):
+        # The eager backend has its OWN _check_structure / _extract_param_buffers copy
+        # (in _EAGER_DRIVER_SOURCE), distinct from the cached/inlined inductor paths the
```
Comment (aorenste) [inline test/test_precompile.py]: [cleanup] This "distinct path" rationale doesn't hold for the shipped code. The test above (`test_wrong_param_count_model_rejected`) uses the default backend, which is `eager` (`backend="inductor"` raises NotImplementedError), so it already runs through this same `_EAGER_DRIVER_SOURCE` `_check_structure` -- there is no separate "cached/inlined inductor path" today. Both tests are worth keeping, but what's distinct is the *input* (count mismatch vs same-count/different-name), not the code path/backend. Suggested reword:

```
# The eager driver's _check_structure rejects a same-param-COUNT but
# different-NAME model (differently-named submodules) rather than silently
# running the traced graph with the wrong weights (invariant 2).
```

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

[cleanup] Several comments were written against an "inductor is the default backend" model, but the shipped default (and only implemented) backend is `eager` -- `backend="inductor"` raises NotImplementedError in torch/_precompile.py. These comments mis-describe the path the test actually exercises:
- L110 (`test_self_contained_exec_needs_no_cache`): "The kernels JIT-compile from the inlined source" -- the eager backend has no kernels; the graph is interpreted and the cache is always empty (artifact=None).
- L123 (`test_wrong_param_count_model_rejected`): "the cached (default) load path" -- the default eager load path execs `python_code`; the eager cache carries no artifact, so nothing is cached.
- L148 (`test_unserializable_in_spec_still_compiles`): "must still compile/run on the inductor backend" -- this call uses the default eager backend.
- L261 (`test_effectful_op_unsupported`): "make the Inductor artifact non-saveable" -- eager rejects here via `_assert_supported` detecting the effect in the captured graph, not via an Inductor save.
- L377 / L389 ("like the inductor backend") are milder but likewise reference a backend that isn't implemented yet.

[cleanup] Coverage gap: the "user input received a gradient" guard in torch/_precompile.py (the `flat_fn` closure in `_capture`, raising "a user input received a gradient") has no test, though its sibling "buffer received a gradient" guard does (`test_buffer_requiring_grad_rejected`). `precompile(lambda x: (x * x).sum().backward(), torch.randn(4, requires_grad=True))` exercises it. The PR description claims coverage of "every enforced invariant."

[cleanup] Coverage gap: the OUT_SPEC hard-failure in `_build_metadata_section` ("precompile cannot serialize the output structure of fn") is untested -- only the IN_SPEC degrade-to-None path is covered (`test_unserializable_in_spec_still_compiles` and the `_context_` variant). Unlike IN_SPEC, OUT_SPEC cannot degrade to None, so an `fn` that *returns* an unregistered `collections.namedtuple` should raise PrecompileError; a test would lock in that non-degrading behavior.
